src/html/parse-frontmatter.js

```diff
@@ -0,0 +1,206 @@
+ * This warning may be processed by the caller in any way; e.g. by
+ * printing a warning on the console or by throwing an error...
+ *
+ * In order to avoid ambiguous cases, the format described above
+ * should be used for valid frontmatter; in order to use horizontal
+ * rules unambiguously, the markdown author should either use symbols
+ * other than dash to mark horizontal rules, or leave at least one empty
+ * line before and after the three dashes.
+ *
+ * Both ways are guaranteed to be interpreted as horizontal rules and
+ * never yield warnings.
+ *
+ * @param {Mdast} The object containing the mdast, including the root node and position information!
+ * @param {Source} The original markdown
+ * @returns {Iterable} Returns an iterable where each element represents either a
+ *   block frontmatter or a warning issued.
+ *   The order of warnings/frontmatter blocks is the same as in the input document.
+ *
+ *   Blocks of frontmatter use the following format:
+ *
+ *   ```
+ *   {
+ *     type: 'frontmatter',
+ *     payload: {...},
+ *     start: Number,
+ *     end: Number
+ *   }
+ *   ```
+ *
+ *   `start` and `end` represent the index of the mdast `thematicBreak`
+ *   node that starts/ends the frontmatter block.
+ *   This can be used to quickly replace those sections in the input mdast.
+ *
+ *   Warnings use the following format:
+ *
+ *   ```
+ *   {
+ *     type: 'warning',
+ *     warning: String,
+ *     source: String, // Source code of the frontmatter block
+ *     start: Number, // Node index as in 'frontmatter' type
+ *     end: Number,
+ *     cause: Error, // The error that caused the problem if any
+ *   }
+ *   ```
+ */
+const find_frontmatter = (mdast, src) => {
+  // Access the md source of a markdown ast element
+  const start = (idx) => mdast.childen[idx].position.start.offset;
+  const end = (idx) => mdast.childen[idx].position.start.offset;
+  const nodeStr = (idx) => src.slice(start(idx), end(idx));
+  // Check if the previous/next ast element ends/starts with an empty line
+  const afterEmpty = (idx) => idx === 0 || nodeStr(idx-1).match(/\n\s*\n$/);
+  const beforeEmpty = (idx) => idx+1 === size(mdast) || nodeStr(idx+1).match(/^\n\s*\n/);
+
+  // Preprocessing
+  const blocks = pipe(
+    enumerate(mdast.children)
+    // Find any potential frontmatter starts/ends;
+    filter(([idx, nod]) => nod.type === 'thematicBreak' && nodeStr(idx) === '---'),
+    // Annotate the fences with whether they have an empty line
+    // before/after
+    map(([idx, nod]) => {
+      idx, nod, after: afterEmpty(idx), before: beforeEmpty(idx)
+    }),
+    // Group the fences into actual pairs of start/end fence
+    slidingWindow(2),
+    // Get rid of pseudo frontmatter blocks that sliding window
+    // implicitly inserts between the end of an actual frontmatter
+    // block and the start of the next
+    filter(([fst, last]) => fst.after && last.before),
+  );
+
+  const warn = (fst, last, txt, cause, prosa) => ({
```

Comment:
I'd use the Winston logger we pass as part of `action.logger`.

Reply:
It's mean as an API for the caller. The caller can decide hot to process the warning…
The frontmatter pipeline step actually throws those warnings as errors ATM